### PR TITLE
web: switch from `crypto` to `SubtleCrypto`

### PR DESCRIPTION
**Pull request by valerybugakov**
## Context

It's a preparation for the Webpack 5 upgrade: https://github.com/sourcegraph/sourcegraph/pull/22580.

Webpack 5 [removed](https://webpack.js.org/blog/2020-10-10-webpack-5-release/#automatic-nodejs-polyfills-removed) automatic Node.js polyfills. To avoid adding polyfills to our bundle, we can switch to the client implementation of the SHA-256 algorithm. That's the only case when Node.js polyfill, specifically `crypto`, is required for our web app.

## Changes

- Switched from `crypto` to the `SubtleCrypto` API for SHA-256 algorithm.

### CODE REVIEW

**Review by eseliger — APPROVED** [review]
LGTM. I hope this module doesn't blow up the bundle size.

**Comment by valerybugakov** [comment]
Added info about bundle size to the description 👍

**Review by vovakulikov — APPROVED** [review]
LGTM. I think 8kb it's fine but if It would be a problem and we need only sha256 we could come up with our own client implementation or get some vendor code and add it to our codebase like https://geraintluff.github.io/sha256/

**Comment by pdubroy** [comment]
Dumb question — do we need a library at all, or can we rely on the browser APIs for this? https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/digest seems to be supported pretty widely.

**Review by abeatrix — APPROVED** [review]

**Comment by valerybugakov** [comment]
Hey @pdubroy, you're right. Initially, I looked into the `SubtleCrypto` API and found that it's not supported in Microsoft Edge that we aim to support. But after looking into the [compatibility table](https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/digest#browser_compatibility) more closely, I see that only SHA-1 is not supported, which means we can rely on the native API. I've updated the PR.

I also took a look into our production bundle, and here's the difference:

1. With `crypto` usage:

Screenshot 2021-07-08 at 17 44 47

2. Without `crypto` polyfill:

Screenshot 2021-07-09 at 08 03 23

**Review by vovakulikov — APPROVED** [review]

File: client/shared/src/util/hashCode.ts [hunk 1]
```diff
@@ -1,14 +1,16 @@
...
  * The hash algorithm here is using Base64 encoding of a SHA256 hash
  *
  */
+export async function hashCode(input: string): Promise<string> {
+    // See `SubtleCrypto` API docs:
+    // https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/digest#converting_a_digest_to_a_hex_string
+    const messageUint8 = new TextEncoder().encode(input) // encode as (utf-8) Uint8Array
+    const hashBuffer = await crypto.subtle.digest('SHA-256', messageUint8) // hash the message
+    const hashArray = [...new Uint8Array(hashBuffer)] // convert buffer to byte array
+    const base64Hash = btoa(String.fromCharCode(...hashArray))
```
Comment (pdubroy) [inline client/shared/src/util/hashCode.ts:13]:
```suggestion
    return btoa(String.fromCharCode(...hashArray))
```

File: client/shared/src/util/hashCode.ts [hunk 2]
```diff
@@ -1,14 +1,16 @@
...
 
-export function hashCode(input: string): string {
-    const hash = createHash('sha256').update(input).digest('base64')
-
-    return hash
+    return base64Hash
```
Comment (pdubroy) [inline client/shared/src/util/hashCode.ts:15]:
```suggestion
```

File: client/shared/src/util/hashCode.ts [hunk 3]
```diff
@@ -1,14 +1,16 @@
...
 /**
  * Returns a hash code value for a string.
  *
  * The hash algorithm here is using Base64 encoding of a SHA256 hash
  *
  */
+export async function hashCode(input: string): Promise<string> {
+    // See `SubtleCrypto` API docs:
+    // https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/digest#converting_a_digest_to_a_hex_string
+    const messageUint8 = new TextEncoder().encode(input) // encode as (utf-8) Uint8Array
```
Comment (pdubroy) [inline client/shared/src/util/hashCode.ts:10]: Nit: Consider dropping the trailing comments on these lines, as they don't add much.

**Review by pdubroy — APPROVED** [review]
nice!

**Pull request merged**